src/gui_fb.c

```diff
@@ -0,0 +1,874 @@
+static void blit_srccopy(Drawable *src, const Rect *srect, Drawable *dst, const Rect *drect);
+
+#if DEBUG
+#define unassert(a)   if (!(a)) unassert_handler(#a,__FILE__, __LINE__)
+
+/* exit graphics mode for error message */
+static void unassert_handler(char *msg, char *file, int line)
+{
+    host_gui_deinit();
+    printf("Assertion failed: %s at %s:%d\n", msg, file, line);
+    exit(255);
+}
+
+static void catch_signals(int signo)
+{
+    host_gui_deinit();
+    printf("SIGNAL %d\n", signo);
+    exit(1);
+}
+#else
+#define unassert(a)
+#endif
+
+void host_gui_begin_composite(void)
+{
+}
+
+void host_gui_end_composite(void)
+{
+}
+
+void host_gui_flush(const Rect *r)
+{
+    Rect cr;
+    cr.x = 0;
+    cr.y = 0;
+    cr.w = fb.width;
+    cr.h = fb.height;
+    unassert(r);
+    blit_srccopy(&bb, &cr, &fb, &cr);
+}
+
+void host_gui_resize(uint64_t w, uint64_t h)
+{
+}
+
+/* return rect adjusted to current clip rectangle */
+static Rect *get_clip_rect(const Rect *rect)
+{
+    static Rect r;
+
+    r = *rect;
+    if (r.w < 1 || r.h < 1) return NULL;
+    r.w += r.x;
+    r.h += r.y;
+    if (r.x >= clip.w || r.y >= clip.h) return NULL;
+    if (r.w <= clip.x || r.h <= clip.y) return NULL;
+    if (clip.x > r.x) r.x = clip.x;
+    if (clip.y > r.y) r.y = clip.y;
+    if (r.w > clip.w) r.w = clip.w;
+    if (r.h > clip.h) r.h = clip.h;
+    r.w -= r.x;
+    r.h -= r.y;
+    return &r;
+}
+
+/* set global clipping rectangle */
+void host_gui_set_clip(const Rect *rect)
+{
+    /* store as x, y, x1, y1 */
+    if (rect) {
+        clip = *rect;
+        clip.w += clip.x;
+        clip.h += clip.y;
+        if (clip.x < 0) clip.x = 0;
+        if (clip.y < 0) clip.y = 0;
+        if (clip.w >= fb.width) clip.w = fb.width - 1;
+        if (clip.h >= fb.height) clip.h = fb.height - 1;
+    } else {
+        clip.x = 0;
+        clip.w = fb.width - 1;
+        clip.y = 0;
+        clip.h = fb.height - 1;
+    }
+}
+
+/* set color for Drawables */
+void host_gui_set_color(uint8_t r, uint8_t g, uint8_t b, uint8_t a)
+{
+    //FIXME premul colors with alpha ?
+    draw_color = (a << 24) + (r << 16) + (g << 8)  + b;
+}
+
+void host_gui_texture_color(Texture *texture, uint8_t r, uint8_t g, uint8_t b)
+{
+    texture->r = r;
+    texture->g = g;
+    texture->b = b;
+    texture->color = (r << 16) + (g << 8)  + b;
+}
+
+/* allocate drawable for passed data and return a handle to it */
+Texture *host_gui_create_texture(void *data, uint64_t width, uint64_t height, uint64_t pitch, uint64_t mode)
+{
+    Texture *t;
+
+    t = malloc(sizeof(Texture));
+    unassert(t);
+    t->pixtype = fb.pixtype;
+    t->bpp = fb.bpp;
+    t->bytespp = fb.bytespp;
+    t->width = width;
+    t->height = height;
+    t->pitch = pitch;
+    t->size = height * t->pitch;
+    t->pixels = malloc(t->size);
+    unassert(t->pixels);
+    t->r = t->g = t->b = 0xff;
+    t->color = 0xffffff;
+    memcpy(t->pixels, data, t->size);
+    return t;
+}
+
```

Comment:
In thwe raw driver example code I'm considering switching to a single alloc for doing this ! I'll do that today and push it.

Reply:
That's a good idea, considering how many textures may be created and destroyed during each screen refresh. Will update.

Reply:
Done, single alloc used for textures.